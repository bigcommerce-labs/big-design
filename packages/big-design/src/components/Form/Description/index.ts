export {
  FormControlDescription,
  type FormControlDescriptionProps,
  type FormControlDescriptionLinkProps,
} from './Description';
