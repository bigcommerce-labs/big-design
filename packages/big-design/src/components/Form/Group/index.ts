export { FormGroup, FormGroupContext, type FormGroupProps } from './Group';
