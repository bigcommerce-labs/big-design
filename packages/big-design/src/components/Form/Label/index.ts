export { FormControlLabel, type FormControlLabelProps } from './Label';
