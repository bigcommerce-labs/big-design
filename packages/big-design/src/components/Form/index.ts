export { Form, type FormProps } from './Form';
export { FormControlDescription, type FormControlDescriptionLinkProps } from './Description';
export { FormControlError } from './Error';
export { FormGroup, type FormGroupProps } from './Group';
export { FormControlLabel, type FormControlLabelProps } from './Label';
