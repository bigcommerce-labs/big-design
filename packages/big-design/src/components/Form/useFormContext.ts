import { createContext, useContext } from 'react';

interface FormContextProps {
  fullWidth?: boolean;
}

export const FormContext = createContext<FormContextProps>({});

export const useFormContext = () => {
  const context = useContext(FormContext);

  if (!context) {
    throw new Error('useFormContext must be used within a FormContext');
  }

  return context;
};
