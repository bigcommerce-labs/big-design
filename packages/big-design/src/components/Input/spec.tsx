import { AddIcon } from '@bigcommerce/big-design-icons';
import { fireEvent, render } from '@testing-library/react';
import 'jest-styled-components';
import React from 'react';

import { Form } from '../Form';

import { Input } from './index';

test('forwards ref', () => {
  const ref = React.createRef<HTMLInputElement>();
  const { container } = render(<Input ref={ref} />);
  const input = container.querySelector('input');

  expect(ref.current).toBe(input);
});

test('renders an input tag', () => {
  const { container } = render(<Input />);

  expect(container.querySelector('input')).toBeInTheDocument();
});

test('renders an input with matched label', () => {
  const { queryByLabelText } = render(<Input label="Test Label" />);

  // This one checks for matching id and htmlFor
  expect(queryByLabelText('Test Label')).toBeInTheDocument();
});

test('create unique ids if not provided', () => {
  const { queryByTestId } = render(
    <>
      <Input label="Test Label" data-testid="item1" />
      <Input label="Test Label" data-testid="item2" />
    </>,
  );

  const item1 = queryByTestId('item1') as HTMLInputElement;
  const item2 = queryByTestId('item2') as HTMLInputElement;

  expect(item1).toBeDefined();
  expect(item2).toBeDefined();
  expect(item1.id).not.toBe(item2.id);
});

test('respects provided id', () => {
  const { container } = render(<Input id="test" label="Test Label" />);
  const input = container.querySelector('#test') as HTMLInputElement;

  expect(input.id).toBe('test');
});

test('matches label htmlFor with id provided', () => {
  const { container } = render(<Input id="test" label="Test Label" />);
  const label = container.querySelector('label') as HTMLLabelElement;

  expect(label.htmlFor).toBe('test');
});

test('renders a description', () => {
  const descriptionText = 'This is a description';
  const { queryByText } = render(<Input description={descriptionText} />);

  expect(queryByText(descriptionText)).toBeInTheDocument();
});

test('renders an error', () => {
  const errorText = 'This is an error';
  const { queryByText } = render(
    <Form.Group>
      <Input error={errorText} />
    </Form.Group>,
  );

  expect(queryByText(errorText)).toBeInTheDocument();
});

test('accepts a Label Component', () => {
  const CustomLabel = (
    <Input.Label>
      This is a custom Label
      <a href="#" data-testid="test">
        has a url
      </a>
    </Input.Label>
  );

  const { queryByTestId } = render(<Input label={CustomLabel} />);

  expect(queryByTestId('test')).toBeInTheDocument();
});

test('does not accept non-Label Components', () => {
  const NotALabel = (
    <div>
      This is a not custom Label Component
      <a href="#" data-testid="test">
        has a url
      </a>
    </div>
  );

  const { queryByTestId } = render(<Input label={NotALabel} />);

  expect(queryByTestId('test')).not.toBeInTheDocument();
});

test('accepts a Description Component', () => {
  const CustomDescription = (
    <Input.Description>
      This is a custom Description
      <a href="#" data-testid="test">
        has a url
      </a>
    </Input.Description>
  );

  const { queryByTestId } = render(<Input description={CustomDescription} />);

  expect(queryByTestId('test')).toBeInTheDocument();
});

test('does not accept non-Description Components', () => {
  const NotADescription = (
    <div>
      This is a not custom description
      <a href="#" data-testid="test">
        has a url
      </a>
    </div>
  );

  const { queryByTestId } = render(<Input description={NotADescription} />);

  expect(queryByTestId('test')).not.toBeInTheDocument();
});

test('accepts an Error Component', () => {
  const CustomError = (
    <Input.Error>
      This is a custom Error Component
      <a href="#" data-testid="test">
        has a url
      </a>
    </Input.Error>
  );

  const { queryByTestId } = render(
    <Form.Group>
      <Input error={CustomError} />
    </Form.Group>,
  );

  expect(queryByTestId('test')).toBeInTheDocument();
});

test('does not accept non-Error Components', () => {
  const NotAnError = (
    <div>
      This is a not a custom error component
      <a href="#" data-testid="test">
        has a url
      </a>
    </div>
  );

  const { queryByTestId } = render(
    <Form.Group>
      <Input error={NotAnError} />
    </Form.Group>,
  );

  expect(queryByTestId('test')).not.toBeInTheDocument();
});

test('renders iconLeft', () => {
  const { queryByTestId } = render(<Input iconLeft={<AddIcon data-testid="icon" />} />);

  expect(queryByTestId('icon')).toBeInTheDocument();
});

test('renders iconRight', () => {
  const { queryByTestId } = render(<Input iconRight={<AddIcon data-testid="icon" />} />);

  expect(queryByTestId('icon')).toBeInTheDocument();
});

test('renders both icons', () => {
  const { queryByTestId } = render(
    <Input iconRight={<AddIcon data-testid="icon-right" />} iconLeft={<AddIcon data-testid="icon-left" />} />,
  );

  expect(queryByTestId('icon-left')).toBeInTheDocument();
  expect(queryByTestId('icon-right')).toBeInTheDocument();
});

test('renders all together', () => {
  const { container } = render(
    <Input
      label="This is a label"
      description="This is a description"
      iconRight={<AddIcon data-testid="icon-right" />}
      iconLeft={<AddIcon data-testid="icon-left" />}
    />,
  );

  expect(container.firstChild).toMatchSnapshot();
});

test('is invalid when required and has no value', () => {
  const { getByLabelText } = render(<Input label="Label" description="This is a description" required />);

  const input = getByLabelText('Label') as HTMLInputElement;

  expect(input.checkValidity()).toEqual(false);
});

test('is valid when required and has a value', () => {
  const { getByLabelText } = render(
    <Input label="Label" value="Hello world" description="This is a description" required />,
  );

  const input = getByLabelText('Label') as HTMLInputElement;

  expect(input.checkValidity()).toEqual(true);
});

test('it is invalid when it was an error', () => {
  const { getByLabelText } = render(
    <Input label="Label" error="Some error" description="This is a description" required />,
  );
  const input = getByLabelText('Label') as HTMLInputElement;

  expect(input.checkValidity()).toEqual(false);

  fireEvent.change(input, { target: { value: 'Some new value ' } });

  expect(input.checkValidity()).toEqual(false);
});

test('error shows with valid string', () => {
  const error = 'Error';
  const { container, rerender } = render(
    <Form.Group>
      <Input error="" />
    </Form.Group>,
  );

  expect(container.querySelector('[class*="StyledError"]')).not.toBeInTheDocument();

  rerender(
    <Form.Group>
      <Input error={error} />
    </Form.Group>,
  );

  expect(container.querySelector('[class*="StyledError"]')).toBeInTheDocument();
});
