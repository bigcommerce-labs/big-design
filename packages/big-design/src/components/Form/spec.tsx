import { fireEvent, render } from '@test/utils';
import 'jest-styled-components';
import React from 'react';

import { Fieldset } from '../Fieldset';
import { Input } from '../Input';

import { Form, FormGroup } from './index';

test('forwards ref', () => {
  const ref = React.createRef<HTMLFormElement>();

  const { container } = render(<Form ref={ref} />);
  const form = container.querySelector('form');

  expect(form).toBe(ref.current);
});

test('calls onSubmit', () => {
  const onSubmit = jest.fn();
  const { container } = render(<Form onSubmit={onSubmit} />);
  const form = container.querySelector('form') as HTMLFormElement;

  fireEvent.submit(form);

  expect(onSubmit).toHaveBeenCalled();
});

test('simple form render', () => {
  const { container } = render(
    <Form>
      <Fieldset
        legend="Primary contact"
        description="Minim velit quis aute adipisicing adipisicing do do exercitation cupidatat enim ex voluptate consequat labore."
      >
        <FormGroup>
          <Input
            label="First Name"
            description="This is an example description for First Name"
            placeholder="Placeholder text"
          />
        </FormGroup>

        <FormGroup>
          <Input
            label="Middle Name"
            description="This is an example description for Last Name. Featuring a Left Icon."
            placeholder="Placeholder text"
          />
        </FormGroup>
      </Fieldset>
    </Form>,
  );

  expect(container.firstChild).toMatchSnapshot();
});
