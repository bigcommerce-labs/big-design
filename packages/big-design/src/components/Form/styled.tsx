import { theme as defaultTheme } from '@bigcommerce/big-design-theme';
import styled from 'styled-components';

import { withMargins } from '../../helpers';
import { StyledFileUploaderWrapper } from '../FileUploader/styled';
import { StyledInputWrapper } from '../Input/styled';
import { StyledTextareaWrapper } from '../Textarea/styled';

import { FormProps } from './Form';

export const StyledForm = styled.form<FormProps>`
  ${withMargins()}

  ${({ theme }) => theme.breakpoints.tablet} {
    ${StyledInputWrapper},
    ${StyledTextareaWrapper},
    ${StyledFileUploaderWrapper} {
      max-width: ${({ fullWidth, theme }) => (fullWidth ? '100%' : theme.helpers.remCalc(416))};
    }
  }
`;

StyledForm.defaultProps = { theme: defaultTheme };
