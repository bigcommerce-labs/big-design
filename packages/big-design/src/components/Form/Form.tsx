import hoistNonReactStatics from 'hoist-non-react-statics';
import React, { Ref } from 'react';

import { StyledForm } from './styled';
import { Error as FormError } from './Error';
import { Fieldset } from './Fieldset';
import { Label } from './Label';
import { Row } from './Row';

interface PrivateProps {
  forwardedRef: Ref<HTMLFormElement>;
}

type FormProps = React.FormHTMLAttributes<HTMLFormElement>;

class StyleableForm extends React.PureComponent<PrivateProps & FormProps> {
  static Label = Label;
  static Error = FormError;
  static Fieldset = Fieldset;
  static Row = Row;

  render() {
    const { forwardedRef, ...props } = this.props;

    return <StyledForm {...props} ref={forwardedRef} />;
  }
}

const FormWithForwardedRef = React.forwardRef<HTMLFormElement, FormProps>(({ className, style, ...props }, ref) => (
  <StyleableForm {...props} forwardedRef={ref} />
));

export const Form = hoistNonReactStatics(FormWithForwardedRef, StyleableForm);
