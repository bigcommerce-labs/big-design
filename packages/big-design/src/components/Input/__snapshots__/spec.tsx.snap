// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders all together 1`] = `
.c5 {
  vertical-align: middle;
  height: 1.5rem;
  width: 1.5rem;
}

.c1 {
  color: #313440;
  margin: 0 0 1rem;
  color: #5E637A;
  font-size: 0.875rem;
  font-weight: 400;
  line-height: 1.25rem;
  margin: 0 0 0.75rem;
}

.c1:last-child {
  margin-bottom: 0;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #FFFFFF;
  border-radius: 0.25rem;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  min-height: 2.25rem;
  padding: 0.25rem 0.75rem;
  width: 100%;
  border: 1px solid #D9DCE9;
}

.c3:hover:not([disabled]) {
  border: 1px solid #B4BAD1;
}

.c3[disabled] {
  background-color: #ECEEF5;
}

.c7 {
  border: 0;
  box-sizing: border-box;
  color: #313440;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 1.5rem;
  margin-top: 0.25rem;
  padding: 0;
  width: 100%;
  padding-right: 0.25rem;
  padding-left: 0.25rem;
}

.c7:focus {
  outline: none;
}

.c7::-webkit-input-placeholder {
  color: #B4BAD1;
  font-size: 1rem;
}

.c7::-moz-placeholder {
  color: #B4BAD1;
  font-size: 1rem;
}

.c7:-ms-input-placeholder {
  color: #B4BAD1;
  font-size: 1rem;
}

.c7::placeholder {
  color: #B4BAD1;
  font-size: 1rem;
}

.c7[disabled] {
  background-color: #ECEEF5;
}

.c4 {
  color: #5E637A;
  -webkit-flex: 0 0 1.5rem;
  -ms-flex: 0 0 1.5rem;
  flex: 0 0 1.5rem;
  height: 1.5rem;
}

.c6 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 100%;
  margin-top: -0.25rem;
}

.c0 {
  color: #313440;
  margin: 0 0 1rem;
  font-size: 1rem;
  font-weight: 600;
  line-height: 1.5rem;
  margin: 0 0 0.5rem;
  display: inline-block;
  margin-bottom: 0.25rem;
}

.c0::after {
  color: #5E637A;
  content: ' (optional)';
  font-weight: 400;
}

@media (min-width:720px) {
  .eLjSIS .c2,
  .eLjSIS .styled__StyledTextareaWrapper-c1uos0-0 {
    max-width: 26rem;
  }
}

<div>
  <label
    class="c0"
    for="input_0"
  >
    This is a label
  </label>
  <p
    class="c1"
  >
    This is a description
  </p>
  <span
    class="c2 c3"
  >
    <div
      class="c4"
    >
      <svg
        class="c5"
        data-testid="icon-left"
        fill="currentColor"
        height="24"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="24"
      >
        <title />
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M18 13h-5v5c0 .55-.45 1-1 1s-1-.45-1-1v-5H6c-.55 0-1-.45-1-1s.45-1 1-1h5V6c0-.55.45-1 1-1s1 .45 1 1v5h5c.55 0 1 .45 1 1s-.45 1-1 1z"
        />
      </svg>
    </div>
    <div
      class="c6"
    >
      <input
        class="c7"
        id="input_0"
      />
    </div>
    <div
      class="c4"
    >
      <svg
        class="c5"
        data-testid="icon-right"
        fill="currentColor"
        height="24"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="24"
      >
        <title />
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M18 13h-5v5c0 .55-.45 1-1 1s-1-.45-1-1v-5H6c-.55 0-1-.45-1-1s.45-1 1-1h5V6c0-.55.45-1 1-1s1 .45 1 1v5h5c.55 0 1 .45 1 1s-.45 1-1 1z"
        />
      </svg>
    </div>
  </span>
</div>
`;
